File: src/lib/tmux.js
'use strict';

const childProcess = require('child_process');

function sessionExists(sessionName) {
  try {
    childProcess.execFileSync('tmux', ['has-session', '-t', sessionName], {
      stdio: 'ignore',
    });
    return true;
  } catch (_e) {
    return false;
  }
}

function sendKeys(sessionName, text, targetPane) {
  const target = targetPane || sessionName;

  // Clear any existing unsubmitted text on the line first
  childProcess.execFileSync('tmux', ['send-keys', '-t', target, 'C-u']);

  // -l sends literal text (no tmux key interpretation)
  // -- prevents text starting with - from being parsed as flags
  childProcess.execFileSync('tmux', [
    'send-keys',
    '-t',
    target,
    '-l',
    '--',
    text,
  ]);

  // Send Enter as a separate keystroke
  childProcess.execFileSync('tmux', ['send-keys', '-t', target, 'Enter']);
}

function capturePane(sessionName, lines = 50, targetPane) {
  const target = targetPane || sessionName;

  return childProcess.execFileSync(
    'tmux',
    ['capture-pane', '-t', target, '-p', '-S', `-${lines}`],
    { encoding: 'utf8' }
  );
}

function sendInterrupt(sessionName, targetPane) {
  const target = targetPane || sessionName;
  childProcess.execFileSync('tmux', ['send-keys', '-t', target, 'C-c']);
}

function sendEscape(sessionName, targetPane) {
  const target = targetPane || sessionName;
  childProcess.execFileSync('tmux', ['send-keys', '-t', target, 'Escape']);
}

function sendArrowDown(sessionName, count, targetPane) {
  const target = targetPane || sessionName;
  for (let i = 0; i < count; i++) {
    childProcess.execFileSync('tmux', ['send-keys', '-t', target, 'Down']);
  }
}

function sendEnter(sessionName, targetPane) {
  const target = targetPane || sessionName;
  childProcess.execFileSync('tmux', ['send-keys', '-t', target, 'Enter']);
}

function renameSession(oldName, newName) {
  childProcess.execFileSync('tmux', ['rename-session', '-t', oldName, newName]);
}

module.exports = {
  sessionExists,
  sendKeys,
  capturePane,
  sendInterrupt,
  sendEscape,
  sendArrowDown,
  sendEnter,
  renameSession,
};

File: internal/tmux/tmux.go
package tmux

import (
	"fmt"
	"os/exec"
	"time"
)

func SessionExists(name string) bool {
	cmd := exec.Command("tmux", "has-session", "-t", name)
	return cmd.Run() == nil
}

func SendKeys(sessionName, text, targetPane string) error {
	target := sessionName
	if targetPane != "" {
		target = targetPane
	}
	// Clear line first
	exec.Command("tmux", "send-keys", "-t", target, "C-u").Run()
	// Send text literally
	if err := exec.Command("tmux", "send-keys", "-t", target, "-l", "--", text).Run(); err != nil {
		return err
	}
	// Send Enter
	return exec.Command("tmux", "send-keys", "-t", target, "Enter").Run()
}

func CapturePane(sessionName string, lines int, targetPane string) (string, error) {
	target := sessionName
	if targetPane != "" {
		target = targetPane
	}
	out, err := exec.Command("tmux", "capture-pane", "-t", target, "-p", "-S", fmt.Sprintf("-%d", lines)).Output()
	if err != nil {
		return "", err
	}
	return string(out), nil
}

func SendInterrupt(sessionName, targetPane string) error {
	target := sessionName
	if targetPane != "" {
		target = targetPane
	}
	return exec.Command("tmux", "send-keys", "-t", target, "C-c").Run()
}

func SendEscape(sessionName, targetPane string) error {
	target := sessionName
	if targetPane != "" {
		target = targetPane
	}
	return exec.Command("tmux", "send-keys", "-t", target, "Escape").Run()
}

func SendArrowDown(sessionName string, count int, targetPane string) error {
	target := sessionName
	if targetPane != "" {
		target = targetPane
	}
	for i := 0; i < count; i++ {
		if err := exec.Command("tmux", "send-keys", "-t", target, "Down").Run(); err != nil {
			return err
		}
		time.Sleep(50 * time.Millisecond)
	}
	return nil
}

func SendEnter(sessionName, targetPane string) error {
	target := sessionName
	if targetPane != "" {
		target = targetPane
	}
	return exec.Command("tmux", "send-keys", "-t", target, "Enter").Run()
}

func RenameSession(oldName, newName string) error {
	return exec.Command("tmux", "rename-session", "-t", oldName, newName).Run()
}
